### PR TITLE
Dynamic web content filters

### PR DESCRIPTION
**Pull request by virlatinus**
| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/234
| Related developer documentation PR URL | N/A
| Issues addressed (#s or URLs) | N/A
| BC breaks? | N/A
| Deprecations? | N/A

#### Description:
This PR adds a new feature to Dynamic Web Content items: The ability to configure them to be "campaign or filter" based.
Right now, DWC items are only campaign based, but with this new feature the users will be able to configure filters (like in segments) so a Landing Page with a DWC token can show content based on specific conditions.

NOTES:
- This PR fixes an issue with pages increasing in size with each save because the CSS classes to the HTML tag were getting duplicated. 
- This PR also adds a label with the slot type to the slots headers when hovering over them in the builders.

#### Steps to test this PR:
1. Apply this PR
2. Create or edit a Dynamic Web Content item.
3. Notice the "Is campaign based" toggle button.

image

4. When No is selected, the `Slot Name` field is available and the filters tab is displayed so new filters can be created.

image

5. On a landing page, add a token or a slot for the new dynamic web content you just created.

image

6. When using DWC slots, you need to provide the requested slot name.

image

7. Visit the page and if the conditions on the filters are met, the DWC will be displayed.

image

8. To test a DWC item with filters in a third party page, you can use the following code:

```
<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <title>It works</title>
  </head>
  <body>
    <h1>DWC Test</h1>
<hr/>

<div data-slot="dwc" data-param-slot-name="myslot">
  <h1>Dynamic web content for myslot</h1>
</div>

<script>
    (function(w,d,t,u,n,a,m){w['MauticTrackingObject']=n;
        w[n]=w[n]||function(){(w[n].q=w[n].q||[]).push(arguments)},a=d.createElement(t),
        m=d.getElementsByTagName(t)[0];a.async=1;a.src=u;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//mautic/index_dev.php/mtc.js','mt');

    mt('send', 'pageview');
</script>

  </body>
</html>
```

9. Visit the page and the DWC slot will be replaced if the filters match:

image

10. Make sure the current campaign-based functionality is not affected. See The [Dynamic Web Content documentation](https://mautic.org/docs/en/dwc/index.html) for details on how to test this scenario.

### CODE REVIEW

File: app/bundles/CoreBundle/Assets/js/libraries/froala/plugins/token.js [hunk 1]
```diff
@@ -32,89 +32,102 @@
...
+                        tokens[k] = nv;
+                    } else if (k.match(/pagelink=/i) && v.match(/a:/)){
+                        delete tokens[k];
+                        nv = v.replace('a:', '');
+                        k = '<a title=\'Page Link\' href=\'' + k + '\'>' + nv + '</a>';
+                        tokens[k] = nv;
+                    } else if (k.match(/dwc=/i) && v.match(/a:/)){
+                        delete tokens[k];
+                        nv = v.replace('a:', '');
+                        k = '<a title=\'DYnamic Web Content\' href=\'' + k + '\'>' + nv + '</a>';
```
Comment (dongilbert) [inline app/bundles/CoreBundle/Assets/js/libraries/froala/plugins/token.js]: Update the spelling here.

File: app/bundles/CoreBundle/Assets/js/4.builder.js [hunk 2]
```diff
@@ -515,10 +521,8 @@ Mautic.sanitizeHtmlBeforeSave = function(htmlContent) {
     // Remove the slot focus highlight
     htmlContent.find('[data-slot-focus], [data-section-focus]').remove();
 
-    customHtml = Mautic.domToString(htmlContent);
```
Comment (alanhartless) [inline app/bundles/CoreBundle/Assets/js/4.builder.js]: This was added by @escopecz for the Apply button function. Will removing this cause issues?

File: app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php [hunk 3]
```diff
@@ -181,17 +291,63 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             );
         }
 
+        $filterModalTransformer = new FieldFilterTransformer($this->translator);
+        $builder->add(
+            $builder->create(
+                'filters',
```
Comment (alanhartless) [inline app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php]: This seems to be pulling in a bunch of segment filters that DWC doesn't support such as Goto, email bounced, page hits, etc.

File: app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php [hunk 4]
```diff
@@ -27,14 +44,92 @@ class DynamicContentType extends AbstractType
...
+     * @param UserHelper          $userHelper
+     *
+     * @throws \InvalidArgumentException
      */
-    public function __construct(EntityManager $entityManager)
+    public function __construct(EntityManager $entityManager, ListModel $listModel, StageModel $stageModel, TranslatorInterface $translator, LeadModel $leadModel, CategoryModel $categoryModel, EmailModel $emailModel, CorePermissions $security, UserHelper $userHelper)
     {
         $this->em = $entityManager;
+
+        $this->fieldChoices = $listModel->getChoiceFields();
```
Comment (alanhartless) [inline app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php]: This is a lot of duplicated code to ListType.

File: app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php [hunk 5]
```diff
@@ -73,6 +168,19 @@ public function buildForm(FormBuilderInterface $builder, array $options)
         $builder->add('isPublished', 'yesno_button_group');
 
         $builder->add(
+            'isCampaignBased',
```
Comment (alanhartless) [inline app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php]: If not campaign based, how is this DWC item requested without a slot name? Say if I have on my wordpress site three areas that I want content served? Currently campaign DWC is requested by a slot name. If no campaign DWC is found, how does Mautic know what to serve back for the requested slot?

File: app/bundles/DynamicContentBundle/Views/DynamicContent/form.html.php [hunk 6]
```diff
@@ -9,35 +9,188 @@
  * @license     GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
  */
 
-$view->extend('MauticCoreBundle:FormTheme:form_simple.html.php');
+$view->extend('MauticCoreBundle:Default:content.html.php');
 $view->addGlobal('translationBase', 'mautic.dynamicContent');
 $view->addGlobal('mauticContent', 'dynamicContent');
+
```
Comment (alanhartless) [inline app/bundles/DynamicContentBundle/Views/DynamicContent/form.html.php]: Lots of duplicate code to what's in the LeadBundle for lists. Can we not leverage templates  in order to not duplicate this code?

File: app/bundles/LeadBundle/Assets/js/lead.js [hunk 7]
```diff
@@ -282,7 +282,13 @@ Mautic.leadlistOnLoad = function(container) {
         Mautic.activateSearchAutocomplete('list-search', 'lead.list');
     }
 
-    if (mQuery('#leadlist_filters').length) {
+    var prefix = 'leadlist';
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Assets/js/lead.js]: Something in this code is causing both segment filters and DWC filters to throw JS errors around reordering filters. ```Uncaught ReferenceError: el is not defined
    at Object.Mautic.reorderSegmentFilters (app/bundles/LeadBundle/Assets/js/lead.js:359:25)```

**Comment by virlatinus** [comment]
@alanhartless the changes you requested are ready. The PR is ready for testing

File: app/bundles/CoreBundle/Assets/js/4.builder.js (hunk 2 above)
Comment (virlatinus) [reply app/bundles/CoreBundle/Assets/js/4.builder.js]: I restored the line

File: app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php (hunk 5 above)
Comment (virlatinus) [reply app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php]: This is fixed in the latest code. A slot name is used

File: app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php (hunk 3 above)
Comment (virlatinus) [reply app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php]: The filter list was trimmed down

File: app/bundles/DynamicContentBundle/Views/DynamicContent/form.html.php (hunk 6 above)
Comment (virlatinus) [reply app/bundles/DynamicContentBundle/Views/DynamicContent/form.html.php]: There are a lot of differences between the two that's why I used a new one

File: app/bundles/LeadBundle/Assets/js/lead.js (hunk 7 above)
Comment (virlatinus) [reply app/bundles/LeadBundle/Assets/js/lead.js]: I fixed the issue: There was an undefined variable `em` that was causing the trouble

File: app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php [hunk 8]
```diff
@@ -181,22 +256,91 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+        $view->vars['deviceTypes']  = $this->deviceTypesChoices;
+        $view->vars['deviceBrands'] = $this->deviceBrandsChoices;
+        $view->vars['deviceOs']     = $this->deviceOsChoices;
+        $view->vars['tags']         = $this->tagChoices;
+        $view->vars['locales']      = $this->localeChoices;
+    }
+
+    private function filterFieldChoices()
+    {
+        $this->fieldChoices['lead'] = array_filter($this->fieldChoices['lead'], function ($key) {
```
Comment (alanhartless) [inline app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php]: This should be an inclusive list to return only the filters supported rather than filtering out those we don't because otherwise filters injected by plugins will still show or every time we add a new filter, we'll have to add it here.

File: app/bundles/LeadBundle/Model/ListModel.php [hunk 9]
```diff
@@ -586,64 +591,56 @@ public function getChoiceFields()
                 ),
                 'object' => 'lead',
             ],
-            'source_id' => [
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Model/ListModel.php]: Why was this removed?

File: app/bundles/PageBundle/EventListener/BuilderSubscriber.php [hunk 10]
```diff
@@ -250,7 +269,15 @@ public function onPageBuild(Events\PageBuilderEvent $event)
                 'video-camera',
                 'MauticCoreBundle:Slots:gatedvideo.html.php',
                 GatedVideoType::class,
-                600
+                300
+            );
+            $event->addSlotType(
+                'dwc',
+                'Dynamic Content',
```
Comment (alanhartless) [inline app/bundles/PageBundle/EventListener/BuilderSubscriber.php]: We should probably be translating all these slot names. Didn't realize we weren't.

File: app/bundles/DynamicContentBundle/Entity/DynamicContent.php [hunk 11]
```diff
@@ -137,6 +149,17 @@ public static function loadMetadata(ORM\ClassMetadata $metadata)
...
         self::addTranslationMetadata($builder, self::class);
         self::addVariantMetadata($builder, self::class);
+        self::addFiltersMetadata($builder);
+
+        $builder->createField('isCampaignBased', 'boolean')
+                ->columnName('is_campaign_based')
+                ->option('default', 1)
+                ->build();
+
+        $builder->createField('slotName', 'string')
```
Comment (alanhartless) [inline app/bundles/DynamicContentBundle/Entity/DynamicContent.php]: Let's add an index on this (be sure to add it to the migration file as well)

File: app/bundles/DynamicContentBundle/Entity/DynamicContent.php [hunk 12]
```diff
@@ -137,6 +149,17 @@ public static function loadMetadata(ORM\ClassMetadata $metadata)
 
         self::addTranslationMetadata($builder, self::class);
         self::addVariantMetadata($builder, self::class);
+        self::addFiltersMetadata($builder);
+
+        $builder->createField('isCampaignBased', 'boolean')
```
Comment (alanhartless) [inline app/bundles/DynamicContentBundle/Entity/DynamicContent.php]: Let's add an index on this (be sure to add it to the migration file as well)

File: app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php [hunk 13]
```diff
@@ -56,6 +103,19 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ]
         );
 
+        $builder->add(
+            'slotName',
+            'text',
+            [
+                'label'      => 'mautic.dynamicContent.send.slot_name',
```
Comment (alanhartless) [inline app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php]: We need a callback constraint on this to validate not blank if isCampaignBased = false

File: app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php (hunk 8 above)
Comment (alanhartless) [reply app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php]: Also, let's exclude company fields. This feature is for visitors and we won't know their company at that point.

File: app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php (hunk 13 above)
Comment (alanhartless) [reply app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php]: We also need a callback constraint on filters to invalidate the field if filters is empty when isCampaignBased = false

**Comment by alanhartless** [comment]
I can't add filters unless I refresh the page. I don't think they are getting activated or something when ajax loads the form. Also if I'm on the Filters tab and disable is campaign based, I'm left on the Filters page but the tab disappears.

File: app/bundles/DynamicContentBundle/Entity/DynamicContent.php (hunk 12 above)
Comment (virlatinus) [reply app/bundles/DynamicContentBundle/Entity/DynamicContent.php]: Done in latest commit

File: app/bundles/DynamicContentBundle/Entity/DynamicContent.php (hunk 11 above)
Comment (virlatinus) [reply app/bundles/DynamicContentBundle/Entity/DynamicContent.php]: Done in latest commit

**Comment by virlatinus** [comment]
@alanhartless fixed the filters tab issue and the list of filters not being activated.

File: app/bundles/DynamicContentBundle/EventListener/DynamicContentSubscriber.php [hunk 14]
```diff
@@ -174,4 +202,55 @@ public function onTokenReplacement(MauticEvents\TokenReplacementEvent $event)
...
+            $slot     = $divContent->item($i);
+            $slotName = $slot->getAttribute('data-param-slot-name');
+            $dwcs     = $this->dynamicContentHelper->getDwcsBySlotName($slotName);
+            /** @var DynamicContent $dwc */
+            foreach ($dwcs as $dwc) {
+                if ($dwc->getIsCampaignBased()) {
+                    continue;
+                }
+                if ($lead && $this->matchFilterForLead($dwc->getFilters(), $leadArray)) {
+                    $slotContent = $lead ? $this->dynamicContentHelper->getRealDynamicContent($dwc->getName(), $lead, $dwc) : '';
```
Comment (alanhartless) [inline app/bundles/DynamicContentBundle/EventListener/DynamicContentSubscriber.php]: Should be $dwc->getSlotName()

File: app/bundles/DynamicContentBundle/Helper/DynamicContentHelper.php [hunk 15]
```diff
@@ -93,6 +83,34 @@ public function getDynamicContentForLead($slot, $lead)
...
+            $leadArray = $lead->getProfileFields();
+        }
+        $dwcs = $this->getDwcsBySlotName($slotName);
+        /** @var DynamicContent $dwc */
+        foreach ($dwcs as $dwc) {
+            if ($dwc->getIsCampaignBased()) {
+                continue;
+            }
+            if ($lead && $this->matchFilterForLead($dwc->getFilters(), $leadArray)) {
+                $slotContent = $lead ? $this->getRealDynamicContent($dwc->getName(), $lead, $dwc) : '';
```
Comment (alanhartless) [inline app/bundles/DynamicContentBundle/Helper/DynamicContentHelper.php]: Should be $dwc->getSlotName()

File: app/bundles/DynamicContentBundle/Helper/DynamicContentHelper.php [hunk 16]
```diff
@@ -118,4 +136,90 @@ public function replaceTokensInContent($content, $lead)
...
+                }
+
+                $dwcs = $this->getDwcsBySlotName($slotName);
+
+                /** @var DynamicContent $dwc */
+                foreach ($dwcs as $dwc) {
+                    if ($dwc->getIsCampaignBased()) {
+                        continue;
+                    }
+                    $content                   = $lead ? $this->getRealDynamicContent($dwc->getName(), $lead, $dwc) : '';
```
Comment (alanhartless) [inline app/bundles/DynamicContentBundle/Helper/DynamicContentHelper.php]: Should be $dwc->getSlotName()

File: app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php (hunk 5 above)
Comment (alanhartless) [reply app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php]: Fixed

File: app/bundles/DynamicContentBundle/Entity/DynamicContent.php (hunk 12 above)
Comment (alanhartless) [reply app/bundles/DynamicContentBundle/Entity/DynamicContent.php]: Fixed

File: app/bundles/DynamicContentBundle/Entity/DynamicContent.php (hunk 11 above)
Comment (alanhartless) [reply app/bundles/DynamicContentBundle/Entity/DynamicContent.php]: Fixed

File: app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php (hunk 13 above)
Comment (virlatinus) [reply app/bundles/DynamicContentBundle/Form/Type/DynamicContentType.php]: Added on latest commit

**Comment by virlatinus** [comment]
@alanhartless, the new column , the Lifecycle callback and the information in the details page is ready

**Comment by alanhartless** [comment]
Good work - I think this is ready now.

File: app/bundles/DynamicContentBundle/Helper/DynamicContentHelper.php [hunk 17]
```diff
@@ -118,4 +137,110 @@ public function replaceTokensInContent($content, $lead)
...
+                $dwcs = $this->getDwcsBySlotName($slotName);
+
+                /** @var DynamicContent $dwc */
+                foreach ($dwcs as $dwc) {
+                    if ($dwc->getIsCampaignBased()) {
+                        continue;
+                    }
+                    $content                   = $lead ? $this->getRealDynamicContent($dwc->getSlotName(), $lead, $dwc) : '';
+                    $tokens[$token]['content'] = $content;
+                    $tokens[$token]['filters'] = $dwc->getFilters();
```
Comment (Maxell92) [inline app/bundles/DynamicContentBundle/Helper/DynamicContentHelper.php]: It is better to use structured object instead of array:
```
tokens = new Token(); // for line 151
tokens->add($content, $dwc->getFilters());

/***/
class Token {
private $items;

public function add($content, xx $filter) {
$this->items[] = [...];
}
..
}
```
This is the easiest implementation, even better is to have another class for content of Token: https://github.com/mautic/mautic/pull/5082/commits/8287eb543bdb79d1c693dac024d72672d5d77201#diff-f18858aaed88e15db191fa7ffa299f45

I had this simple data class and change it to another structure in this commit because I need to hold more data.

File: app/bundles/DynamicContentBundle/Helper/DynamicContentHelper.php [hunk 18]
```diff
@@ -118,4 +137,110 @@ public function replaceTokensInContent($content, $lead)
...
+    }
+
+    /**
+     * @param $slot
+     * @param $lead
+     * @param $dwc
+     *
+     * @return string
+     */
+    public function getRealDynamicContent($slot, $lead, DynamicContent $dwc)
```
Comment (Maxell92) [inline app/bundles/DynamicContentBundle/Helper/DynamicContentHelper.php]: Could we typehint parameters? What is `$slot`? Is it possible to use `Lead $lead`?

File: app/bundles/DynamicContentBundle/Helper/DynamicContentHelper.php [hunk 19]
```diff
@@ -118,4 +137,110 @@ public function replaceTokensInContent($content, $lead)
...
+    }
+
+    /**
+     * @param string $slotName
+     *
+     * @return array|\Doctrine\ORM\Tools\Pagination\Paginator
+     */
+    public function getDwcsBySlotName($slotName)
+    {
+        return $this->dynamicContentModel->getEntities(
```
Comment (Maxell92) [inline app/bundles/DynamicContentBundle/Helper/DynamicContentHelper.php]: It is alway better to encapsulate implementation. This should only call something like `return $this->dynamicContentModel->getEntitiesBySlotName($slotName)` and query should be builded inside of Repository class.

File: app/bundles/DynamicContentBundle/Helper/DynamicContentHelper.php [hunk 20]
```diff
@@ -118,4 +137,110 @@ public function replaceTokensInContent($content, $lead)
...
+            ]
+        );
+    }
+
+    /**
+     * @param Lead $lead
+     *
+     * @return array
+     */
+    public function convertLeadToArray($lead)
```
Comment (Maxell92) [inline app/bundles/DynamicContentBundle/Helper/DynamicContentHelper.php]: This method should be inside of `Lead` class. It depends only on Lead.

File: app/bundles/PageBundle/EventListener/BuilderSubscriber.php [hunk 21]
```diff
@@ -108,9 +115,21 @@ public function onPageBuild(Events\PageBuilderEvent $event)
             $event->addAbTestWinnerCriteria('page.dwelltime', $dwellTime);
         }
 
-        if ($event->tokensRequested([$this->pageTokenRegex])) {
+        if ($event->tokensRequested([$this->pageTokenRegex, $this->dwcTokenRegex])) {
             $event->addTokensFromHelper($tokenHelper, $this->pageTokenRegex, 'title', 'id', false, true);
 
+            // add only filter based dwc tokens
+            $dwcTokenHelper = new BuilderTokenHelper($this->factory, 'dynamicContent');
```
Comment (Maxell92) [inline app/bundles/PageBundle/EventListener/BuilderSubscriber.php]: Why the new code uses `factory`? It should be deprecated?

File: app/bundles/EmailBundle/EventListener/MatchFilterForLeadTrait.php [hunk 22]
```diff
@@ -0,0 +1,187 @@
...
+ */
+trait MatchFilterForLeadTrait
+{
+    /**
+     * @param array $filter
+     * @param array $lead
+     *
+     * @return bool
+     */
+    protected function matchFilterForLead(array $filter, array $lead)
```
Comment (Maxell92) [inline app/bundles/EmailBundle/EventListener/MatchFilterForLeadTrait.php]: Very long and complicated method. Do we have a tests for use cases? It is not part of this PR.

File: app/bundles/DynamicContentBundle/Controller/DynamicContentApiController.php [hunk 23]
```diff
@@ -39,6 +39,9 @@ public function getAction($objectAlias)
     {
         $lead    = $this->getModel('lead')->getCurrentLead();
         $content = $this->get('mautic.helper.dynamicContent')->getDynamicContentForLead($objectAlias, $lead);
+        if (empty($content)) {
```
[the remark this answers is not in the document]
Comment (alanhartless) [reply app/bundles/DynamicContentBundle/Controller/DynamicContentApiController.php]: This is on purpose I think. The function naming could probably be improved. But it first checks for campaign content (getDynamicContentForLead) . If that is empty, it then checks for content based on filters (getDynamicContent*Slot*ForLead).

File: app/bundles/PageBundle/EventListener/BuilderSubscriber.php (hunk 21 above)
Comment (alanhartless) [reply app/bundles/PageBundle/EventListener/BuilderSubscriber.php]: It is; it's because BuilderTokenHelper has been around for a while and still depends on it. We haven't refactored it to be a service.

File: app/bundles/PageBundle/EventListener/BuilderSubscriber.php [hunk 24]
```diff
@@ -108,9 +115,21 @@ public function onPageBuild(Events\PageBuilderEvent $event)
...
+            $expr           = $this->factory->getDatabase()->getExpressionBuilder()->andX('e.is_campaign_based <> 1 and e.slot_name is not null');
```
Comment (alanhartless) [inline app/bundles/PageBundle/EventListener/BuilderSubscriber.php]: Although we shouldn't be using factory here.

File: app/bundles/DynamicContentBundle/Controller/DynamicContentApiController.php (hunk 23 above)
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/DynamicContentBundle/Controller/DynamicContentApiController.php]: You are right, I did not notice this one. The logic should be inside class than and here in Controller. It prevents mistake I have done here :)
